File: classloader-vortrag-javademos/src/main/java/name/felixbecker/talks/classloader/duplicatedloading/DuplicateLoadingDemo.java
package name.felixbecker.talks.classloader.duplicatedloading;

import java.lang.reflect.Method;
import java.nio.file.Files;
import java.nio.file.Path;
import java.nio.file.Paths;

public class DuplicateLoadingDemo {

    private static final SimpleDefiningClassLoader cl1 = new SimpleDefiningClassLoader();
    private static final SimpleDefiningClassLoader cl2 = new SimpleDefiningClassLoader();

    public static void main(String... args) throws Exception {

        Path path = Paths.get("/home/becker/git/classloader-vortrag/classloader-vortrag-javademos/simple-app-manually-compiled/name/felixbecker/SimpleApp.class");

        byte[] data = Files.readAllBytes(path);

        Class<?> cl1Clazz = cl1.defineManually(data);
        Class<?> cl2Clazz = cl2.defineManually(data);

        Object cl1Instance = cl1Clazz.newInstance();
        Thread.sleep(1000);
        Object cl2Instance = cl2Clazz.newInstance();

        Method cl1Method = cl1Clazz.getDeclaredMethod("greet", String.class);
        Method cl2Method = cl2Clazz.getDeclaredMethod("greet", String.class);

        cl1Method.invoke(cl1Instance, "felix");
        cl2Method.invoke(cl2Instance, "felix");

        System.out.println(cl1Clazz.getCanonicalName() + " " + cl1Clazz);
        System.out.println(cl2Clazz.getCanonicalName() + " " + cl2Clazz);


    }




}

File: classloader-vortrag-javademos/src/main/scala/name/felixbecker/talks/classloader/DuplicateLoadingDemo.scala
package name.felixbecker.talks.classloader

import java.nio.file.{Files, Paths}

class SimpleDefiningClassLoader extends ClassLoader {
  def defineManually(bytes: Array[Byte]): Class[_] = defineClass(null, bytes, 0, bytes.length)
}

object DuplicateLoadingDemo extends App {

  val cl1 = new SimpleDefiningClassLoader
  val cl2 = new SimpleDefiningClassLoader

  val path = Paths.get("/home/becker/kaufhof/git/classloader-vortrag/classloader-vortrag-javademos/simple-app-manually-compiled/name/felixbecker/SimpleApp.class")
  val data = Files.readAllBytes(path)

  val cl1Clazz = cl1.defineManually(data)
  val cl2Clazz = cl2.defineManually(data)

  val cl1Instance = cl1Clazz.newInstance
  Thread.sleep(1000)
  val cl2Instance = cl2Clazz.newInstance

  val cl1Method = cl1Clazz.getDeclaredMethod("greet", classOf[String])
  val cl2Method = cl2Clazz.getDeclaredMethod("greet", classOf[String])

  cl1Method.invoke(cl1Instance, "felix")
  cl2Method.invoke(cl2Instance, "felix")

  println(cl1Clazz.getCanonicalName + " " + cl1Clazz)
  println(cl2Clazz.getCanonicalName + " " + cl2Clazz)

  println("Class 1 is assignable from class 1: " + cl1Clazz.isAssignableFrom(cl1Clazz))
  println("Class 1 is assignable from class 2: " + cl1Clazz.isAssignableFrom(cl2Clazz))
}
